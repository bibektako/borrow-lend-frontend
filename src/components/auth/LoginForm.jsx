import React from "react";
import { Formik, Form, Field, ErrorMessage } from "formik";
import * as Yup from "yup";
import { ToastContainer, toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";
import { useNavigate } from "react-router-dom";
import facebook from "../../assets/images/facebook.png";
import google from "../../assets/images/google.png";
import logo from "../../assets/images/logo.png";
import { loginUserApi } from "../../api/authAPI";

const LoginForm = () => {
  const navigate = useNavigate();
  const initialValues = {
    email: "",
    password: "",
    remember: false,
  };

  const validationSchema = Yup.object({
    email: Yup.string().email("Invalid email address").required("Required"),
    password: Yup.string().min(6, "Minimum 6 characters").required("Required"),
  });

  const handleSubmit = async (values, { setSubmitting, resetForm }) => {
    try {
      const response = await loginUserApi({
        email: values.email,
        password: values.password,
      });

      const token = response.data?.token;
      const user = response.data?.data;

      if (!token || !user) {
        toast.error("Invalid login response from the server.");
        setSubmitting(false);
        return;
      }

      // Save token (localStorage or sessionStorage based on "remember" checkbox)
      if (values.remember) {
        localStorage.setItem("token", token);
        localStorage.setItem("user", JSON.stringify(user));
      } else {
        sessionStorage.setItem("token", token);
        sessionStorage.setItem("user", JSON.stringify(user));
      }

      toast.success("Welcome back!");
      resetForm();

      setTimeout(() => {
        setSubmitting(false);

        if (user.role === "admin") {
          navigate("/admin/categories");
        } else {
          navigate("/");
        }
      }, 1500);
    } catch (error) {
      toast.error(
        error.response?.data?.message || "Login failed. Please try again."
      );
      setSubmitting(false);
    }
  };

  return (
    <div className="min-h-screen flex items-center justify-center bg-gray-50">
      <div className="bg-white p-8 rounded-lg shadow-lg max-w-md w-full">
        <ToastContainer />
        <div className="flex justify-center mb-4">
          <img src={logo} alt="logo" className="w-12 h-12" />
        </div>
        <h2 className="text-center text-2xl font-bold text-gray-900">
          Welcome back
        </h2>
        <p className="text-center text-gray-600">
          Sign in to your account to continue
        </p>

        <Formik
          initialValues={initialValues}
          validationSchema={validationSchema}
          onSubmit={handleSubmit}
        >
          {({ isSubmitting }) => (
            <Form className="mt-6 space-y-4">
              <div>
                <label
                  htmlFor="email"
                  className="block text-sm font-medium text-gray-700"
                >
                  Email
                </label>
                <Field
                  name="email"
                  type="email"
                  placeholder="Enter your email"
                  className="mt-1 w-full px-4 py-2 border border-gray-300 rounded-lg focus:ring-blue-500 focus:border-blue-500"
                />
                <ErrorMessage
                  name="email"
                  component="div"
                  className="text-sm text-red-500 mt-1"
                />
              </div>

              <div>
                <label
                  htmlFor="password"
                  className="block text-sm font-medium text-gray-700"
                >
                  Password
                </label>
                <Field
                  name="password"
                  type="password"
                  placeholder="Enter your password"
                  className="mt-1 w-full px-4 py-2 border border-gray-300 rounded-lg focus:ring-blue-500 focus:border-blue-500"
                />
                <ErrorMessage
                  name="password"
                  component="div"
                  className="text-sm text-red-500 mt-1"
                />
                <div className="text-right mt-1">
                  <a href="#" className="text-sm text-blue-600 hover:underline">
                    Forgot password?
                  </a>
                </div>
              </div>

              <div className="flex items-center">
                <Field type="checkbox" name="remember" className="mr-2" />
                <label htmlFor="remember" className="text-sm text-gray-700">
                  Remember me
                </label>
              </div>
              <button
                type="submit"
                disabled={isSubmitting}
                className="w-full bg-blue-600 text-white py-2 rounded-lg hover:bg-blue-700 transition"
              >
                {isSubmitting ? "Signing in..." : "Sign In"}
              </button>
            </Form>
          )}
        </Formik>

        <div className="flex items-center my-4">
          <div className="flex-grow h-px bg-gray-300" />
          <span className="mx-2 text-sm text-gray-500">OR CONTINUE WITH</span>
          <div className="flex-grow h-px bg-gray-300" />
        </div>
        <div className="flex space-x-4">
          <button className="w-1/2 border py-2 rounded-lg flex items-center justify-center hover:bg-gray-100">
            <img src={google} alt="Google" className="w-5 h-5 mr-2" />
            Google
          </button>
          <button className="w-1/2 border py-2 rounded-lg flex items-center justify-center hover:bg-gray-100">
            <img src={facebook} alt="Facebook" className="w-5 h-5 mr-2" />
            Facebook
          </button>
        </div>

        <p className="mt-4 text-center text-sm text-gray-600">
          Donâ€™t have an account?{" "}
          <a
            href="#"
            className="text-blue-600 hover:underline"
            onClick={() => navigate("/signUp")}
          >
            Register
          </a>
        </p>
      </div>
    </div>
  );
};

export default LoginForm;
