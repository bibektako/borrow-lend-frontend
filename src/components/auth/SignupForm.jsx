import React from "react";

import { Formik, Form, Field, ErrorMessage } from "formik";
import * as Yup from "yup";
import { toast, ToastContainer } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";
import { useNavigate } from "react-router-dom";
import facebook from '../../assets/images/facebook.png';
import google from '../../assets/images/google.png';
import { registerUserApi } from '/src/api/authAPI.js';


const SignupForm = () => {
  const navigate = useNavigate();

  const initialValues = {
    fullName: "",
    email: "",
    phoneNo: "",
    password: "",
    terms: false,
  };

  const validationSchema = Yup.object({
    fullName: Yup.string().required("Full name is required"),
    email: Yup.string()
      .email("Invalid email address")
      .required("Email is required"),
    phoneNo: Yup.string()
      .matches(/^[0-9]{10}$/, "Phone number must be 10 digits")
      .required("Phone number is required"),
    password: Yup.string()
      .min(8, "Password must be at least 8 characters")
      .matches(/[0-9]/, "Password must contain a number")
      .matches(/[!@#$%^&*]/, "Password must contain a special character")
      .required("Password is required"),
    terms: Yup.boolean().oneOf(
      [true],
      "You must accept the terms and privacy policy"
    ),
  });

  const handleSubmit = async (values, { resetForm, setSubmitting }) => {
    try {
      const response = await registerUserApi({
        username: values.fullName,
        email: values.email,
        phone: values.phoneNo,
        password: values.password,
      });

      toast.success("Account created successfully!");
      resetForm();

      setTimeout(() => {
        navigate("/signin"); // Redirect after signup
      }, 1500);
    } catch (error) {
      toast.error(
        error.response?.data?.message || "Signup failed. Please try again."
      );
    } finally {
      setSubmitting(false);
    }
  };

  return (
    <div className="min-h-screen flex items-center justify-center bg-gray-50">
      <div className="bg-white p-8 rounded-lg shadow-lg max-w-md w-full">
        <ToastContainer />
        <h2 className="text-center text-2xl font-bold text-gray-900">
          Create an account
        </h2>
        <p className="text-center text-gray-600 mb-4">
          Join our community to start lending and borrowing
        </p>

        <Formik
          initialValues={initialValues}
          validationSchema={validationSchema}
          onSubmit={handleSubmit}
        >
          {({ isSubmitting }) => (
            <Form className="space-y-4">
              <div>
                <label className="block text-sm font-medium text-gray-700">
                  Full Name
                </label>
                <Field
                  name="fullName"
                  placeholder="Enter your full name"
                  className="w-full mt-1 px-4 py-2 border border-gray-300 rounded-lg"
                />
                <ErrorMessage
                  name="fullName"
                  component="div"
                  className="text-sm text-red-500 mt-1"
                />
              </div>

              <div>
                <label className="block text-sm font-medium text-gray-700">
                  Email
                </label>
                <Field
                  name="email"
                  type="email"
                  placeholder="Enter your email"
                  className="w-full mt-1 px-4 py-2 border border-gray-300 rounded-lg"
                />
                <ErrorMessage
                  name="email"
                  component="div"
                  className="text-sm text-red-500 mt-1"
                />
              </div>

              <div>
                <label className="block text-sm font-medium text-gray-700">
                  Phone Number
                </label>
                <Field
                  name="phoneNo"
                  type="tel"
                  placeholder="Enter your phone number"
                  className="w-full mt-1 px-4 py-2 border border-gray-300 rounded-lg"
                />
                <ErrorMessage
                  name="phoneNo"
                  component="div"
                  className="text-sm text-red-500 mt-1"
                />
              </div>

              <div>
                <label className="block text-sm font-medium text-gray-700">
                  Password
                </label>
                <Field
                  name="password"
                  type="password"
                  placeholder="Create a password"
                  className="w-full mt-1 px-4 py-2 border border-gray-300 rounded-lg"
                />
                <ErrorMessage
                  name="password"
                  component="div"
                  className="text-sm text-red-500 mt-1"
                />
                <p className="text-xs text-gray-500 mt-1">
                  Password must be at least 8 characters long with a number and
                  a special character.
                </p>
              </div>

              <div className="flex items-center">
                <Field type="checkbox" name="terms" className="mr-2" />
                <label className="text-sm text-gray-700">
                  I agree to the{" "}
                  <a href="#" className="text-blue-600 hover:underline">
                    Terms of Service
                  </a>{" "}
                  and{" "}
                  <a href="#" className="text-blue-600 hover:underline">
                    Privacy Policy
                  </a>
                </label>
              </div>
              <ErrorMessage
                name="terms"
                component="div"
                className="text-sm text-red-500 mt-1"
              />

              <button
                type="submit"
                disabled={isSubmitting}
                className="w-full bg-blue-600 text-white py-2 rounded-lg hover:bg-blue-700 transition"
              >
                {isSubmitting ? "Registering..." : "Register"}
              </button>
            </Form>
          )}
        </Formik>

        <div className="flex items-center my-4">
          <div className="flex-grow h-px bg-gray-300" />
          <span className="mx-2 text-sm text-gray-500">OR REGISTER WITH</span>
          <div className="flex-grow h-px bg-gray-300" />
        </div>

        <div className="flex space-x-4">
          <button className="w-1/2 border py-2 rounded-lg flex items-center justify-center hover:bg-gray-100">
            <img src={google} alt="Google" className="w-5 h-5 mr-2" />
            Google
          </button>
          <button className="w-1/2 border py-2 rounded-lg flex items-center justify-center hover:bg-gray-100">
            <img
              src={facebook}
              alt="Facebook"
              className="w-5 h-5 mr-2"
            />
            Facebook
          </button>
        </div>

        <p className="mt-4 text-center text-sm text-gray-600">
          Already have an account?{" "}
          <span
            className="text-blue-600 hover:underline cursor-pointer"
            onClick={() => navigate("/signin")}
          >
            Sign in
          </span>
        </p>
      </div>
    </div>
  );
};

export default SignupForm;
