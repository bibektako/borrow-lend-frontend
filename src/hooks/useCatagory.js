import { useQuery } from "@tanstack/react-query";

import { getAllCategoryService } from "../services/admin/categoryService";


export const useCategories = () => {
  const {
    data: categoryData, // We get the raw response here, let's call it categoryData
    isLoading,
    isError,
    error,
  } = useQuery({
    queryKey: ["categories"],
    
    queryFn: getAllCategoryService,
    
    staleTime: 1000 * 60 * 15,
  });

  const categories = categoryData?.data || [];

  return {
    categories, // The clean array of category objects
    isLoading,  // True while fetching
    isError,    // True if an error occurred
    error,      // The error object
  };
};