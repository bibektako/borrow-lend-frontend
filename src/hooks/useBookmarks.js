import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';
import { useContext } from 'react';
import { AuthContext } from '../auth/Authprovider';
import { toast } from 'react-hot-toast';

import {
  getBookmarksApi,
  addBookmarkApi,
  removeBookmarkApi,
} from '../api/authAPI'; // Make sure this path is correct

export const useBookmarks = () => {
  const { user } = useContext(AuthContext);
  const queryClient = useQueryClient();

  const { data: bookmarks = [], isLoading } = useQuery({
    queryKey: ['bookmarks', user?._id],
    queryFn: async () => {
      const response = await getBookmarksApi();
      return response.data.data; 
    },
    enabled: !!user,
  });

  const isBookmarked = (itemId) => {
    return bookmarks.some((bookmark) => bookmark._id === itemId);
  };

  const { mutate: toggleBookmark, isPending: isToggling } = useMutation({
    mutationFn: async ({ itemId, isCurrentlyBookmarked }) => {
      if (isCurrentlyBookmarked) {
        return removeBookmarkApi(itemId);
      } else {
        return addBookmarkApi(itemId);
      }
    },
    onSuccess: (response, variables) => {
      const message = variables.isCurrentlyBookmarked
        ? "Removed from wishlist"
        : "Added to wishlist";
      toast.success(message);

      queryClient.invalidateQueries({ queryKey: ['bookmarks', user?._id] });
    },
    onError: (error) => {
      toast.error(error.response?.data?.message || "Could not update wishlist.");
    },
  });

  const handleToggleBookmark = (itemId) => {
    const isCurrentlyBookmarked = isBookmarked(itemId);
    toggleBookmark({ itemId, isCurrentlyBookmarked });
  };

  return {
    bookmarks,
    isLoading,
    toggleBookmark: handleToggleBookmark, 
    isToggling,
    isBookmarked,
    bookmarkCount: bookmarks.length,
  };
};