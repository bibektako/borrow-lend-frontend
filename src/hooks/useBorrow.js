import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import {
  createBorrowRequestService,
  getBorrowRequestsService,
  updateBorrowRequestService,
} from "../services/borrowService";
import { toast } from "react-toastify";
import { useContext } from "react";
import { AuthContext } from "../auth/Authprovider";

export const useBorrowRequests = () => {
  const { token } = useContext(AuthContext);
  const { data, isLoading, error, isError, refetch} = useQuery({
    queryKey: ["borrow-requests"],
    queryFn: () => getBorrowRequestsService(token),
    enabled: !!token,
  });
  return {
    requests: data?.data || [],
    isLoading,
    error,
    isError,
    refetch
  };
};

export const useCreateBorrowRequest = () => {
  const queryClient = useQueryClient();
  const { token } = useContext(AuthContext);

  return useMutation({
    mutationFn: (itemId) => createBorrowRequestService(itemId, token),
    onSuccess: (data) => {
      toast.success(data?.message || "Borrow request sent successfully!");
      queryClient.invalidateQueries({ queryKey: ["items"] });
      queryClient.invalidateQueries({ queryKey: ["item"] });
      queryClient.invalidateQueries({ queryKey: ["borrow-requests"] });
    },
    onError: (error) => {
      toast.error(error?.message || "Failed to send borrow request.");
    },
  });
};

export const useUpdateBorrowRequest = () => {
  const queryClient = useQueryClient();
  const { token } = useContext(AuthContext);

  return useMutation({
    mutationFn: ({ requestId, status }) =>
      updateBorrowRequestService(requestId, status, token),
    onSuccess: (data) => {
      toast.success(data?.message || "Request updated!");
      queryClient.invalidateQueries({ queryKey: ["borrow-requests"] });
    },
    onError: (error) => {
      toast.error(error?.message || "Failed to update request.");
    },
  });
};
