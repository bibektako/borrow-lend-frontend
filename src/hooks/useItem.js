import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query";
import {
  getItemsService,
  getItemByIdService,
  getMyItemsService,
  createItemService,
  updateItemService,
  deleteItemService,
} from "../services/itemService";
import { getAllCategoryService } from "../services/admin/categoryService";
import { toast } from "react-toastify";
import { useContext } from "react";
import { AuthContext } from "../auth/Authprovider";

export const useItems = (filters = {}) => {
  const { data, isLoading, error, isError } = useQuery({
    queryKey: ["items", filters],
    queryFn: () => getItemsService(filters),
  });

  return {
    items: data?.data || [],
    isLoading,
    error,
    isError,
  };
};

export const useMyItems = () => {
  const { data, isLoading, error, isError } = useQuery({
    queryKey: ["my-items"],
    queryFn: getMyItemsService,
  });
  return {
    items: data?.data || [],
    isLoading,
    error,
    isError,
  };
};
export const useCategories = () => {
    const { data, isLoading, error } = useQuery({
        queryKey: ["categories"],
        queryFn: getAllCategoryService,
    });
    return {
        categories: data?.data?.categories || [],
        isLoadingCategories: isLoading,
        errorCategories: error,
    };
};

export const useItemById = (id) => {
  const { data, isLoading, error, isError } = useQuery({
    queryKey: ["item", id],
    queryFn: () => getItemByIdService(id),
    enabled: !!id,
  });

  return {
    item: data?.data || null,
    isLoading,
    error,
    isError,
  };
};

export const useCreateItem = () => {
  const queryClient = useQueryClient();
  const { token } = useContext(AuthContext);

  return useMutation({
    mutationFn: (formData) => createItemService(formData, token),
    onSuccess: (data) => {
      toast.success(data?.message || "Item created successfully!");
      queryClient.invalidateQueries({ queryKey: ["items"] });
    },
    onError: (error) => {
      toast.error(error?.message || "Failed to create item.");
    },
  });
};

export const useUpdateItem = () => {
  const queryClient = useQueryClient();
  const { token } = useContext(AuthContext);

  return useMutation({
    mutationFn: ({ id, formData }) => updateItemService(id, formData, token),
    onSuccess: (data, variables) => {
      toast.success(data?.message || "Item updated successfully!");

      queryClient.invalidateQueries({ queryKey: ["items"] });
      queryClient.invalidateQueries({ queryKey: ["item", variables.id] });
    },
    onError: (error) => {
      toast.error(error?.message || "Failed to update item.");
    },
  });
};

export const useDeleteItem = () => {
  const queryClient = useQueryClient();
  const { token } = useContext(AuthContext);

  return useMutation({
    mutationFn: (id) => deleteItemService(id, token),
    onSuccess: (data) => {
      toast.success(data?.message || "Item deleted successfully!");

      queryClient.invalidateQueries({ queryKey: ["items"] });
    },
    onError: (error) => {
      toast.error(error?.message || "Failed to delete item.");
    },
  });
};
