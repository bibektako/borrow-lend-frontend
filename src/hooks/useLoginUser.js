import { toast } from "react-toastify";
import { loginUserService } from "../services/authService";
import { useMutation } from "@tanstack/react-query";
import { AuthContext } from "../auth/Authprovider";
import { useContext } from "react";
export const useLoginUser =() => {
    const { login } = useContext(AuthContext)

    return useMutation(
        {
            mutationFn: loginUserService,
            mutationKey: ['login-key'],
            onSuccess: (data) => {
                login(data?.data, data?.token)
                toast.success(data?.message || "Login Success")
            },
            onError: (err)=>{
                toast.error(err?.message || "Login Failed")
            }
        }
    )

}