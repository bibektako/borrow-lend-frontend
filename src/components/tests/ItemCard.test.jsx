// src/components/__tests__/ItemCard.test.jsx
import React from "react";
import { describe, it, beforeEach, expect, vi } from "vitest";
import { render, screen, fireEvent } from "@testing-library/react";
import ItemCard from "../user/ItemsCard";
import { BrowserRouter } from "react-router-dom";
import { AuthContext } from "../../auth/Authprovider";
import { useBookmarks } from "../../hooks/useBookmarks";
import { useCreateBorrowRequest } from "../../hooks/useBorrow";

vi.mock("../../hooks/useBookmarks");
vi.mock("../../hooks/useBorrow");

const mockToggleBookmark = vi.fn();
const mockCreateBorrowRequest = vi.fn();

const mockItem = {
  _id: "item123",
  name: "Sample Item",
  category: { name: "Electronics" },
  borrowingPrice: 150,
  imageUrls: ["test.jpg"],
  owner: { _id: "owner1", location: "Kathmandu" },
  rating: 4.5,
  reviewCount: 12,
  status: "available",
};

describe("SingleItemCard Component", () => {
  beforeEach(() => {
    useBookmarks.mockReturnValue({
      toggleBookmark: mockToggleBookmark,
      isBookmarked: () => false,
      isToggling: false,
    });

    useCreateBorrowRequest.mockReturnValue({
      mutate: mockCreateBorrowRequest,
      isPending: false,
    });
  });

  const renderCard = (user = { _id: "user1" }, item = mockItem) =>
    render(
      <BrowserRouter>
        <AuthContext.Provider value={{ user }}>
          <ItemCard items={[item]} />
        </AuthContext.Provider>
      </BrowserRouter>
    );

  it("renders item details", () => {
    renderCard();
    expect(screen.getByText("Sample Item")).toBeInTheDocument();
    expect(
      screen.getByText(
        (content, element) => element?.textContent === "Rs150.00/day"
      )
    ).toBeInTheDocument();
    expect(screen.getByText(/Electronics/)).toBeInTheDocument();
    expect(screen.getByText(/Kathmandu/)).toBeInTheDocument();
  });

  it("renders star ratings and review count", () => {
    renderCard();
    expect(screen.getByText("(12 reviews)")).toBeInTheDocument();
  });

  it("calls toggleBookmark on heart click", () => {
    renderCard();
    fireEvent.click(screen.getByLabelText("Add to wishlist"));
    expect(mockToggleBookmark).toHaveBeenCalledWith(mockItem._id);
  });

  it("calls borrow request on button click", () => {
    renderCard();
    fireEvent.click(screen.getByTestId("borrow-button"));
    expect(mockCreateBorrowRequest).toHaveBeenCalledWith(mockItem._id);
  });

  it('shows "Your Item" if user is owner', () => {
    renderCard({ _id: "owner1" });
    expect(screen.getByText(/Your Item/)).toBeInTheDocument();
  });

  it("disables borrow button if item is already borrowed", () => {
    const borrowedItem = { ...mockItem, status: "borrowed" };
    renderCard({ _id: "user1" }, borrowedItem);
    expect(screen.getByTestId("borrow-button")).toBeDisabled();
  });

  it("matches snapshot", () => {
    const { asFragment } = renderCard();
    expect(asFragment()).toMatchSnapshot();
  });
});
