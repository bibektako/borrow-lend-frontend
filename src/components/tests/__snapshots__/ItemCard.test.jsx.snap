// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SingleItemCard Component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 gap-6"
  >
    <a
      class="block group w-full"
      data-discover="true"
      href="/item/item123"
    >
      <div
        class="relative flex flex-col h-full bg-white rounded-xl shadow-md overflow-hidden border border-gray-200 transition-all duration-300 hover:shadow-xl hover:-translate-y-1"
      >
        <div
          class="relative h-52 w-full"
        >
          <img
            alt="Sample Item"
            class="w-full h-full object-cover transition-transform duration-500 group-hover:scale-110"
            loading="lazy"
            src="http://localhost:5050/test.jpg"
          />
          <div
            class="absolute top-3 right-3 flex flex-col gap-2"
          >
            <div
              class="inline-flex items-center gap-1.5 px-2.5 py-1 text-xs font-semibold rounded-full bg-green-100 text-green-800"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-circle-check-big"
                fill="none"
                height="14"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="14"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M21.801 10A10 10 0 1 1 17 3.335"
                />
                <path
                  d="m9 11 3 3L22 4"
                />
              </svg>
              Available
            </div>
          </div>
          <button
            aria-label="Add to wishlist"
            class="absolute top-3 left-3 w-9 h-9 flex items-center justify-center rounded-full bg-white/80 backdrop-blur-sm transition-colors duration-200 disabled:opacity-50
                text-gray-600 hover:text-red-500
              "
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-heart"
              fill="none"
              height="18"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="18"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 14c1.49-1.46 3-3.21 3-5.5A5.5 5.5 0 0 0 16.5 3c-1.76 0-3 .5-4.5 2-1.5-1.5-2.74-2-4.5-2A5.5 5.5 0 0 0 2 8.5c0 2.3 1.5 4.05 3 5.5l7 7Z"
              />
            </svg>
          </button>
        </div>
        <div
          class="p-4 flex flex-col flex-grow"
        >
          <p
            class="text-xs font-semibold text-blue-600 uppercase tracking-wider"
          >
            Electronics
          </p>
          <h3
            class="text-md font-bold text-gray-900 mt-1 truncate"
            title="Sample Item"
          >
            Sample Item
          </h3>
          <div
            class="flex items-center text-gray-500 mt-2 text-sm"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-map-pin w-4 h-4 mr-1.5 flex-shrink-0 text-gray-400"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M20 10c0 4.993-5.539 10.193-7.399 11.799a1 1 0 0 1-1.202 0C9.539 20.193 4 14.993 4 10a8 8 0 0 1 16 0"
              />
              <circle
                cx="12"
                cy="10"
                r="3"
              />
            </svg>
            <span
              class="truncate"
            >
              Kathmandu
            </span>
          </div>
          <div
            class="flex items-center mt-3"
          >
            <div
              class="flex items-center"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-star w-4 h-4 text-yellow-400"
                fill="currentColor"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="lucide lucide-star w-4 h-4 text-yellow-400"
                fill="currentColor"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="lucide lucide-star w-4 h-4 text-yellow-400"
                fill="currentColor"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="lucide lucide-star w-4 h-4 text-yellow-400"
                fill="currentColor"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="lucide lucide-star w-4 h-4 text-yellow-400"
                fill="currentColor"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
                />
              </svg>
            </div>
            <p
              class="text-gray-500 ml-2 text-xs"
            >
              (12 reviews)
            </p>
          </div>
          <div
            class="flex-grow"
          />
          <div
            class="flex justify-between items-center pt-4 mt-4 border-t border-gray-100"
          >
            <p
              class="text-lg font-bold text-gray-900"
            >
              150.00
              <span
                class="text-sm font-normal text-gray-500"
              >
                /day
              </span>
            </p>
            <button
              class="px-5 py-2 text-sm font-semibold text-white bg-blue-600 rounded-lg transition-all duration-300 hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-500 disabled:bg-gray-300 disabled:cursor-not-allowed"
              data-testid="borrow-button"
            >
              Borrow
            </button>
          </div>
        </div>
      </div>
    </a>
  </div>
</DocumentFragment>
`;
